docs/messaging/messaging-spans.md

```diff
@@ -196,21 +198,74 @@ Examples:
-| `receive`      | A message is received from a destination by a message consumer/server. |
-| `process`      | A message that was previously received from a destination is processed by a message consumer/server. |
+| `publish`      | One or more messages are provided for publishing to an intermediary. If a single message is published, the context of the "Publish" span can be used as the creation context and no "Create" span needs to be created. |
+| `create`       | A message is created. "Create" spans always refer to a single message and are used to provide a unique creation context for messages in batch publishing scenarios. |
+| `receive`      | One or more messages are requested by a consumer. This operation refers to pull-based scenarios, where consumers explicitly call methods of messaging SDKs to receive messages. |
+| `deliver`      | One or more messages are passed to a consumer. This operation refers to push-based scenarios, where consumer register callbacks which get called by messaging SDKs. |
+
+### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER` if the context of the "Publish" span is used as creation context. |
+| `create`       | `PRODUCER` |
+| `receive`      | `CONSUMER` |
+| `deliver`      | `CONSUMER` |
+
+For cases not covered by the table above, the span kind should be set according
+to the [generic specification about span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind),
+e. g. it should be set to CLIENT for the "Publish" span if its context is not
+used as creation context and if the "Publish" span models a synchronous call to
+the intermediary.
+
+Setting span kinds according to this table ensures that span links between
+consumers and producers always exist between a PRODUCER span on the producer
+side and a CONSUMER span on the consumer side. This allows analysis tools to
+interpret linked traces without the need for additional semantic hints.
+
+### Trace structure
+
+#### Producer spans
+
+"Publish" spans SHOULD be created for operations of providing messages for
+sending or publishing to an intermediary. A single "Publish" span can account
+for a single message, or for multiple messages (in the case of providing
+messages in batches). "Create" spans MAY be created. A single "Create" span
+SHOULD account only for a single message. "Create" spans SHOULD either be
+children or links of the related "Publish" span.
+
+If a "Create" span exists for a message, its context SHOULD be injected into
+the message. If no "Create" span exists and no custom creation context is
+injected into the message, the context of the related "Publish" span SHOULD be
+injected into the message.
```

Comment:
I think we completely lost this point from the OTEP
> A user provides custom creation contexts for the messages that are published, which don't refer to any spans described in this document. This provides flexibility to users to model custom scenarios. In this case, no other additional spans besides the "Publish" span should be created. The "Publish" span should link to the provided creation contexts.

It's also not reflected in the [Context Propagation](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-spans.md#context-propagation) section.

Reply:
@lmolkova What do you think about this wording:

> If a user provides a custom creation context for a message, this context SHOULD be injected into the message.  Otherwise, if a "Create" span exists for a message, its context SHOULD be injected into the message. If no "Create" span exists and no custom creation context is injected into the message, the context of the related "Publish" span SHOULD be injected into the message.

Reply:
The part that's missing that Publish operation should be linked to the message context.

I'd add something g like 

"If a user provides a custom creation context int a message, this context SHOULD NOT be modified: `Create` span  SHOULD NOT be created. The custom creation context should be linked to the `Publish` span."

I.e.:
- when publishing/creating a message:
  - extract context. 
     - if context already exists, add it as a link to Publish span
     - otherwise, start publish/create span, inject context

Reply:
@lmolkova Got it, I ended up changing the wording to this:

> If a user provides a custom creation context in a message, this context SHOULD NOT be modified, a "Create" span SHOULD NOT be created, and the "Publish" span SHOULD link to the custom creation context.  Otherwise, if a "Create" span exists for a message, its context SHOULD be injected into the message. If no "Create" span exists and no custom creation context is injected into the message, the context of the related "Publish" span SHOULD be injected into the message.

Let me know if you think that covers the whole custom creation context scenario.